pkg/controller/podautoscaler/replica_calculator.go

```diff
@@ -449,31 +452,76 @@ func groupPods(pods []*v1.Pod, metrics metricsclient.PodMetricsInfo, resource v1
 	return
 }
 
-func calculatePodRequests(pods []*v1.Pod, container string, resource v1.ResourceName) (map[string]int64, error) {
+// calculateRequests computes the request value for each pod for the specified
+// resource.
+// If container is non-empty, it uses the request of that specific container.
+// If container is empty, it uses pod-level requests if pod-level requests are
+// set on the pod. Otherwise, it sums the requests of all containers in the pod
+// (including restartable init containers).
+// It returns a map of pod names to their calculated request values.
+func calculateRequests(pods []*v1.Pod, container string, resource v1.ResourceName) (map[string]int64, error) {
+	podLevelResourcesEnabled := feature.DefaultFeatureGate.Enabled(features.PodLevelResources)
 	requests := make(map[string]int64, len(pods))
 	for _, pod := range pods {
-		podSum := int64(0)
-		// Calculate all regular containers and restartable init containers requests.
-		containers := append([]v1.Container{}, pod.Spec.Containers...)
-		for _, c := range pod.Spec.InitContainers {
-			if c.RestartPolicy != nil && *c.RestartPolicy == v1.ContainerRestartPolicyAlways {
-				containers = append(containers, c)
-			}
+		var request int64
+		var err error
+		// Determine if we should use pod-level requests: see KEP-2837
+		// https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2837-pod-level-resource-spec/README.md
+		usePodLevelRequests := podLevelResourcesEnabled &&
+			resourcehelpers.IsPodLevelRequestsSet(pod) &&
+			// If a container name is specified, it does not make sense to
```

Comment:
nit: we could use some stronger language here

"If a container name is specified in the HPA, it takes precedence over the pod-level requests"

(Just trying to think ahead when we clean up the alpha flag checks here and someone might accidentally delete the `&& container == ""` part)

Reply:
That's a good point. Fixed. Thanks!